include/download_client.h

```diff
@@ -0,0 +1,241 @@
+ *  - disconnect from the server
+ *  - receive asynchronous event notification on the status of
+ *    download
+ *
+ * Currently, only the HTTP protocol is supported for download.
+ */
+
+#ifndef DOWNLOAD_CLIENT_H__
+#define DOWNLOAD_CLIENT_H__
+
+#include <zephyr/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct download_client;
+
+/** @brief Download states. */
+enum download_client_status {
+	/** Indicates that the client was either never connected
+	 *  to the server or is now disconnected.
+	 */
+	DOWNLOAD_CLIENT_STATUS_IDLE      = 0x00,
+	/** Indicates that the client is connected to the server
+	 *  and there is no ongoing download.
+	 */
+	DOWNLOAD_CLIENT_STATUS_CONNECTED = 0x01,
+	/** Indicates that the client is connected to the server
+	 *  and download is in progress.
+	 */
+	DOWNLOAD_CLIENT_STATUS_DOWNLOAD_INPROGRESS = 0x02,
+	/** Indicates that the client is connected to the server
+	 *  and download is complete.
+	 */
+	DOWNLOAD_CLIENT_STATUS_DOWNLOAD_COMPLETE = 0x03,
+	/** Indicates that the object download is halted by the
+	 *  application. This status indicates that the application
+	 *  identified a failure when handling the
+	 *  @ref DOWNLOAD_CLIENT_EVT_DOWNLOAD_FRAG event.
+	 */
+	DOWNLOAD_CLIENT_STATUS_HALTED = 0x04,
+	/** Indicates that an error occurred and the download
+	 *  cannot continue.
+	 */
+	DOWNLOAD_CLIENT_ERROR = 0xFF
+};
+
+
+/** @brief Download events. */
+enum download_client_evt {
+	/** Indicates an error during download.
+	 *  The application should disconnect and retry the operation
+	 *  when receiving this event.
+	 */
+	DOWNLOAD_CLIENT_EVT_ERROR = 0x00,
+	/** Indicates reception of a fragment during download.
+	 *  The @p fragment field of the @ref download_client object
+	 *  points to the object fragment, and the fragment size
+	 *  indicates the size of the fragment.
+	 */
+	DOWNLOAD_CLIENT_EVT_DOWNLOAD_FRAG = 0x01,
+	/** Indicates that the download is complete.
+	 */
+	DOWNLOAD_CLIENT_EVT_DOWNLOAD_DONE = 0x02,
+};
+
+/** @brief Download client asynchronous event handler.
+ *
+ * The application is notified of the status of the object download
+ * through this event handler.
+ * The application can use the return value to indicate if it handled
+ * the event successfully or not.
+ * This feedback is useful when, for example, a faulty fragment was
+ * received or the application was unable to process a object fragment.
+ *
+ * @param[in] client	The client instance.
+ * @param[in] event     The event.
+ * @param[in] status    Event status (either 0 or an errno value).
+ *
+ * @retval 0 If the event was handled successfully.
+ *           Other values indicate that the application failed to handle
+ *           the event.
+ */
+typedef int (*download_client_event_handler_t)(struct download_client *client,
+				enum download_client_evt event, u32_t status);
+
+/** @brief Object download client instance that describes the state of download.
+ */
+struct download_client {
```

Comment:
A user of this module will typically call the init function with a struct like this one:
```c
struct download_client c= {
	.host = CONFIG_MY_DOWNLOAD_HOST,
	.resource = CONFIG_MY_DOWNLOAD_FILE,
	.callback = my_dfu_event_handler
};
int retval = download_client_init(&c);
```

All other fields of that struct should not be touched.. @oxalistriangularis Since all other fields are "private" fileds, would it be possible to group them under an opaque pointer (and adding one API for allocating the struct)?

Reply:
The fields are neither private nor opaque. The are read-only for the application. The application very much needs access to it to make decisions and advances its procedures.